garageScript/c0d3-app
# Add DOJO exercise submissions to the backend

Changes:
* Adds ability to get and mutate DOJO exercise submissions using GraphQL
* Frontend now gets DOJO exercise submissions from the backend (but it doesn't mutate them on the backend yet, this will be handled in a subsequent pull request)

How to test:
* Create a DOJO module at /admin/lessons/js0/modules or by using GraphQL mutations
* Create DOJO exercises at /curriculum/js0/mentor or by using GraphQL mutations
* Create a exercise submissions by using GraphQL mutations
* Go to the DOJO exercises page at /exercises/js0 and check that the correct exercises are marked as "NOT ANSWERED" or "ANSWERED" depending on how you mutated the exercise submission data
* Notice that if you answer a DOJO exercise at /exercises/js0, it will be marked as "ANSWERED" when you go to the exercise list view, but if you refresh the page, the state will go back to how it was because this pull request doesn't mutate the exercise submissions.

Here are some useful GraphQL queries and mutations for testing (you can run these at /api/graphql):

Add a new module
```graphql
mutation AddModule(
  $lessonId: Int!
  $name: String!
  $content: String!
  $order: Int!
) {
  addModule(
    lessonId: $lessonId
    name: $name
    content: $content
    order: $order
  ) {
    id
  }
}
```
```json
{
  "lessonId": 1,
  "name": "Numbers",
  "content": "We will learn how to use numbers in this lesson.",
  "order": 1
}
```

Add a new exercise to a module
```graphql
mutation AddExercise(
  $moduleId: Int!
  $description: String!
  $answer: String!
  $explanation: String
) {
  addExercise(
    moduleId: $moduleId
    description: $description
    answer: $answer
    explanation: $explanation
  ) {
    id
  }
}
```
```json
{
  "moduleId": 1,
  "description": "```\nlet a = 1\na += 2\n// What is a?\n```",
  "answer": "3",
  "explanation": "`a += 2` is shorthand for `a = a + 2`"
}
```

Add an exercise submission to an exercise
```graphql
mutation AddExerciseSubmission($exerciseId: Int!, $userAnswer: String!) {
  addExerciseSubmission(exerciseId: $exerciseId, userAnswer: $userAnswer) {
    id
    exerciseId
    userId
    userAnswer
  }
}
```
```json
{
  "exerciseId": 1,
  "userAnswer": "3"
}
```

Here is the query that gets run when you go to /exercises/js0
```graphql
query GetExercises {
  lessons {
    title
    docUrl
    slug
  }
  alerts {
    id
    text
    type
    url
    urlCaption
  }
  exercises {
    id
    module {
      name
      lesson {
        slug
      }
    }
    description
    answer
    explanation
  }
  exerciseSubmissions {
    exerciseId
    userAnswer
  }
}
```

This pull request is a part of https://github.com/garageScript/c0d3-app/issues/2251 and https://github.com/garageScript/c0d3-app/issues/2253
